linkerd/http-route/src/http/match/host.rs

```diff
@@ -0,0 +1,205 @@
+use http::Uri;
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub enum MatchHost {
+    Exact(String),
+
+    /// Tokenized reverse list of DNS name suffix labels.
+    ///
+    /// For example: the match `*.example.com` is stored as `["com",
+    /// "example"]`.
+    Suffix(Vec<String>),
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub enum HostMatch {
+    Exact(usize),
+    Suffix(usize),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum InvalidHost {
+    #[error("invalid host: {0}")]
+    Invalid(#[from] url::ParseError),
+
+    #[error("host must not be an IP address")]
+    NumericHost,
+}
+
+// === impl MatchHost ===
+
+impl std::str::FromStr for MatchHost {
+    type Err = InvalidHost;
+
+    fn from_str(host: &str) -> Result<Self, Self::Err> {
+        use url::Host;
+        if let Host::Ipv4(_) | Host::Ipv6(_) = Host::parse(host)? {
+            return Err(InvalidHost::NumericHost);
+        }
+
+        if let Some(host) = host.strip_prefix("*.") {
+            return Ok(Self::Suffix(
+                host.split('.').map(|s| s.to_string()).rev().collect(),
+            ));
+        }
+
+        Ok(Self::Exact(host.to_string()))
+    }
+}
+
+impl MatchHost {
+    pub fn summarize_match(&self, uri: &Uri) -> Option<HostMatch> {
+        let mut host = uri.authority()?.host();
+
+        match self {
+            Self::Exact(h) => {
+                if !h.ends_with('.') {
+                    host = host.strip_suffix('.').unwrap_or(host);
+                }
+                if h == host {
+                    Some(HostMatch::Exact(h.len()))
+                } else {
+                    None
+                }
+            }
+
+            Self::Suffix(suffix) => {
+                if suffix.first().map(|s| &**s) != Some("") {
+                    host = host.strip_suffix('.').unwrap_or(host);
+                }
+                let mut length = 0;
+                for sfx in suffix.iter() {
+                    if !host.ends_with(sfx) {
+                        return None;
+                    }
+                    host = &host[..host.len() - sfx.len()];
```

Comment:
nit, take it or leave it: i think this is equivalent to
```suggestion
                    host = host.strip_suffix(sfx)?;
```